Enforcing valid status transitions

iamconnor4 opened the pull request:
See #26

- Returns 500 error if the file upload has already been confirmed rather than reconfirming.
- Enriched responses for error 500s.

iamconnor4 committed:
Enforcing valid status transitions
- Returns 500 error if the file upload has already been confirmed rather than reconfirming.
- Enriched responses for error 500s.

iamconnor4 committed:
Changed from `error` to `errorType` in `ErrorResponse` as it was failing backend test due to ambigious name

tchoutri commented on server/src/Masna3/Server/Error.hs:
```diff
@@ -1,12 +1,54 @@
+              { errorType = "FileNotFound"
+              , context = Just $ object ["fileId" .= t]
+              }
+      }
```
For "\<something\> not found" errors, we use the [404](https://hackage-content.haskell.org/package/servant-server-0.20.3.0/docs/Servant-Server.html#v:err404) status. You can find a list of the HTTP statuses that are available to us [here](https://http.cat/), although I suggest that we restrict ourselves to:

* 404 when a resource is not found
* 403 when authentication fails (it's not used yet)
* 500 when we have a server-side exception

tchoutri commented on server/src/Masna3/Server/Error.hs:
```diff
@@ -1,12 +1,54 @@
+  , context :: Maybe Value
+  }
+  deriving stock (Generic)
+  deriving anyclass (ToJSON)
```
No need for that. :) 

The intuition is good, but we actually want to have a `context` field that can hold more data than we can actually send back – for security & privacy purposes.

My suggestion is that we can put the metadata of `context` as logging metadata, not in the response. This is done with the [`localData`](https://hackage-content.haskell.org/package/log-base-0.12.1.0/docs/Log-Class.html#v:localData) of our logging framework. 

For instance:

```diff
- Error.throwError (FileNotFound fileId)
+ Log.localData ["file_id" .= fileId] $
+   Error.throwError (FileNotFound fileId)
```

tchoutri reviewed (changes requested):
Good start!

iamconnor4 committed:
Changed `FileNotFound` error code from 500 to 404

iamconnor4 committed:
Removed `ErrorResponse` and we log the metadata of `context` instead.

tchoutri committed:
Rearrange the format of the logs on errors

I have made it so that when serialised, server errors look like this:

```json
  […]
  "data": {
    "error_body": {
      "InvalidTransition": {
        "file_id": "0199c9c6-c550-7dce-951a-2c0e760d1206"
      }
    },
    "error_code": 500
  }
```

The wrappers may look cumbersome but they give us the correct structure
when serialised to JSON.

tchoutri commented:
@iamconnor4 ok I reckon we are near the end. 

For the test, you can copy-paste [this test](https://github.com/tchoutri/masna3/blob/main/server/test/Masna3/Test/File.hs#L28-L36) and add another `void $ assertLeftWithStatus "Confirm File" 500 =<< runRequest (Client.confirmFile result.fileId)`. This will make the test fail (since we are trying to confirm a file that has already been confirmed). Don't forget to edit "test-client-2" in that new test to something else (like `test-client-3`).

iamconnor4 committed:
New test `confirmFileInvalidTransition` which checks if an invalid transition occurs, such as attempting to confirm an already confirmed file.

tchoutri commented:
Perfect, let's get that merged. :)
